File: postgresql-async/src/test/scala/com/github/mauricio/async/db/postgresql/TransactionSpec.scala
package com.github.mauricio.async.db.postgresql

import org.specs2.mutable.Specification
import com.github.mauricio.async.db.util.Log
import scala.concurrent.ExecutionContext.Implicits.global
import com.github.mauricio.async.db.postgresql.exceptions.GenericDatabaseException

class TransactionSpec extends Specification with DatabaseTestHelper {

  val log = Log.get[TransactionSpec]

  val tableCreate = "CREATE TEMP TABLE transaction_test (x integer PRIMARY KEY)"

  def tableInsert(x: Int) = "INSERT INTO transaction_test VALUES (" + x.toString + ")"

  val tableSelect = "SELECT x FROM transaction_test ORDER BY x"

  "transactions" should {

    "commit simple inserts" in {
      withHandler {
        handler =>
          executeDdl(handler, tableCreate)
          await(handler.inTransaction {
            conn =>
              conn.sendQuery(tableInsert(1)).flatMap {
                _ =>
                  conn.sendQuery(tableInsert(2))
              }
          })

          val rows = executeQuery(handler, tableSelect).rows.get
          rows.length === 2
          rows(0)(0) === 1
          rows(1)(0) === 2
      }
    }

    "commit simple inserts with prepared statements" in {
      withHandler {
        handler =>
          executeDdl(handler, tableCreate)
          await(handler.inTransaction {
            conn =>
              conn.sendPreparedStatement(tableInsert(1)).flatMap {
                _ =>
                  conn.sendPreparedStatement(tableInsert(2))
              }
          })

          val rows = executePreparedStatement(handler, tableSelect).rows.get
          rows.length === 2
          rows(0)(0) === 1
          rows(1)(0) === 2
      }
    }

    "rollback on error" in {
      withHandler {
        handler =>
          executeDdl(handler, tableCreate)

          try {
            await(handler.inTransaction {
              conn =>
                conn.sendQuery(tableInsert(1)).flatMap {
                  _ =>
                    conn.sendQuery(tableInsert(1))
                }
            })
            failure("Should not have come here")
          } catch {
            case e: GenericDatabaseException => {
              e.errorMessage.message === "duplicate key value violates unique constraint \"transaction_test_pkey\""
            }
          }

          val rows = executeQuery(handler, tableSelect).rows.get
          rows.length === 0
      }

    }

    "rollback explicitly" in {
      withHandler {
        handler =>
          executeDdl(handler, tableCreate)
          await(handler.inTransaction {
            conn =>
              conn.sendQuery(tableInsert(1)).flatMap {
                _ =>
                  conn.sendQuery("ROLLBACK")
              }
          })

          val rows = executeQuery(handler, tableSelect).rows.get
          rows.length === 0
      }

    }

    "rollback to savepoint" in {
      withHandler {
        handler =>
          executeDdl(handler, tableCreate)
          await(handler.inTransaction {
            conn =>
              conn.sendQuery(tableInsert(1)).flatMap {
                _ =>
                  conn.sendQuery("SAVEPOINT one").flatMap {
                    _ =>
                      conn.sendQuery(tableInsert(2)).flatMap {
                        _ =>
                          conn.sendQuery("ROLLBACK TO SAVEPOINT one")
                      }
                  }
              }
          })

          val rows = executeQuery(handler, tableSelect).rows.get
          rows.length === 1
          rows(0)(0) === 1
      }

    }

  }

}

File: postgresql-async/src/test/java/com/github/aysnc/sql/db/integration/TransactionSpec.kt
package com.github.aysnc.sql.db.integration

import com.github.aysnc.sql.db.verifyException
import com.github.jasync.sql.db.invoke
import com.github.jasync.sql.db.postgresql.exceptions.GenericDatabaseException
import com.github.jasync.sql.db.util.ExecutorServiceUtils
import com.github.jasync.sql.db.util.flatMapAsync
import com.github.jasync.sql.db.util.length
import java.util.concurrent.ExecutionException
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class TransactionSpec : DatabaseTestHelper() {

    private val tableCreate = "CREATE TEMP TABLE transaction_test (x integer PRIMARY KEY)"

    private fun tableInsert(x: Int) = "INSERT INTO transaction_test VALUES ($x)"

    private val tableSelect = "SELECT x FROM transaction_test ORDER BY x"

    @Test
    fun `transactions should commit simple inserts`() {
        withHandler { handler ->
            executeDdl(handler, tableCreate)
            awaitFuture(handler.inTransaction { conn ->
                conn.sendQuery(tableInsert(1)).flatMapAsync(ExecutorServiceUtils.CommonPool) {
                    conn.sendQuery(tableInsert(2))
                }
            })

            val rows = executeQuery(handler, tableSelect).rows
            assertThat(rows.length).isEqualTo(2)
            assertThat(rows(0)(0)).isEqualTo(1)
            assertThat(rows(1)(0)).isEqualTo(2)
        }
    }

    @Test
    fun `transactions should commit simple inserts, prepared statements`() {
        withHandler { handler ->
            executeDdl(handler, tableCreate)
            awaitFuture(handler.inTransaction { conn ->
                conn.sendPreparedStatement(tableInsert(1)).flatMapAsync(ExecutorServiceUtils.CommonPool) {
                    conn.sendPreparedStatement(tableInsert(2))
                }
            })

            val rows = executePreparedStatement(handler, tableSelect).rows
            assertThat(rows.length).isEqualTo(2)
            assertThat(rows(0)(0)).isEqualTo(1)
            assertThat(rows(1)(0)).isEqualTo(2)
        }
    }

    @Test
    fun `transactions should rollback on error`() {
        withHandler { handler ->
            executeDdl(handler, tableCreate)

            val e: GenericDatabaseException =
                verifyException(ExecutionException::class.java, GenericDatabaseException::class.java) {
                    awaitFuture(handler.inTransaction { conn ->
                        conn.sendQuery(tableInsert(1)).flatMapAsync(ExecutorServiceUtils.CommonPool) {
                            conn.sendQuery(tableInsert(1))
                        }
                    })
                } as GenericDatabaseException

            assertThat(e.errorMessage.message).isEqualTo("duplicate key value violates unique constraint \"transaction_test_pkey\"")

            val rows = executeQuery(handler, tableSelect).rows
            assertThat(rows.length).isEqualTo(0)
        }
    }

    @Test
    fun `transactions should rollback explicitly`() {
        withHandler { handler ->
            executeDdl(handler, tableCreate)
            awaitFuture(handler.inTransaction { conn ->
                conn.sendQuery(tableInsert(1)).flatMapAsync(ExecutorServiceUtils.CommonPool) {
                    conn.sendQuery("ROLLBACK")
                }
            })

            val rows = executeQuery(handler, tableSelect).rows
            assertThat(rows.length).isEqualTo(0)
        }
    }

    @Test
    fun `transactions should rollback to savepoint`() {
        withHandler { handler ->
            executeDdl(handler, tableCreate)
            awaitFuture(handler.inTransaction { conn ->
                conn.sendQuery(tableInsert(1)).flatMapAsync(ExecutorServiceUtils.CommonPool) {
                    conn.sendQuery("SAVEPOINT one").flatMapAsync(ExecutorServiceUtils.CommonPool) {
                        conn.sendQuery(tableInsert(2)).flatMapAsync(ExecutorServiceUtils.CommonPool) {
                            conn.sendQuery("ROLLBACK TO SAVEPOINT one")
                        }
                    }
                }
            })

            val rows = executeQuery(handler, tableSelect).rows
            assertThat(rows.length).isEqualTo(1)
            assertThat(rows[0](0)).isEqualTo(1)
        }
    }
}
